### ISSUE TITLE
tubeAmp crash on Ardour plugin analyse 

### ISSUE THREAD

**Issue by brummer10** [olegkapitonov/Kapitonov-Plugins-Pack#4]
Hi

I notice the tubeAmp crash on Ardour plugin analysis, my guess is, it's because the convolver can't handle buffersize changes. I've had the same issue with some guitarix plugs once, that's why I wrote a wrapper for zita-convolver which handle that. Beside that it allow to run plugin analysis on Ardour, it also avoid crash on loop points, and allow smaller buffersizes the 64. 
The related part of the code is here:
https://sourceforge.net/p/guitarix/git/ci/master/tree/trunk/src/LV2/DSP/gx_convolver.cc#l256

EDIT: for stereo it's this part:
https://sourceforge.net/p/guitarix/git/ci/master/tree/trunk/src/LV2/DSP/gx_convolver.cc#l375

The theory is, it fill the convolver buffer since it is full, when the incoming buffer is to small, After processing, it filter out the related part and give back to main process.

regards
hermann

**Comment by olegkapitonov** [comment]
Hi Hermann!

For one year I couldn't understand why this is happening! Thank you!
But I think this is not because buffer size changes, in my code convolver buffer always 64.
I use wrapper too, simpler than yours but it should work with big buffers.

Looks like Ardour creates second instance of the plugin class for the analysis,
and this second object can't load profile data correctly from *tapf, or has
convolver objects uninitialized.
I'm trying to fix this.

**Comment by brummer10** [comment]
You could do a simple test, just start jack with a frame-rate of 2048 (that's the frame rate ardour use for plugin analysis), and watch the analysis. tubeAmp wouldn't crash. 
Ardour didn't start a second instance for analysis, it just change the buffersize to 2048. for the analysis cycles. 

regards
hermann

**Comment by brummer10** [comment]
Okay, I see now, your convolver wrapper handle this very good. !
 I do a deeper check now and found that you've setup the forge_buffer in kpp_tubeamp_gui.c to small. Set it to 4096 will solve this issue. 

regards
hermann

**Comment by olegkapitonov** [comment]
Strange, I still have crashes with forge_buf 4096 and 409600 too.
Crash occurs when input data copy to convolver buffer starts.
Pointer to this buffer is broken at this moment, I don't understand why yet.

**Comment by brummer10** [comment]
Yes, still crash, just not in any case when doing analysis.
So I put a control function on the convolver create and convolver start_process functions in plugin.h. 
Turns out that, start_process for the cabsym fail sometimes, then, it crash.

**Comment by olegkapitonov** [comment]
I fixed problem at least, everything works now.
I am preparing commit.
Problem is very very deep, I will describe it after commit.

**Comment by olegkapitonov** [comment]
I pushed commit to master.

About problem.
I added debug prints to "instantiate" function in plugin,
and I clearly saw that Ardour calls "instantiate" second time
when I click "Plugin Analysis". 

Therefore, Ardour (at least 5.12 which I use) creates
second "analysis" instance of plugin with big buffer 4096.
This instance works in parallel with the first, "sound" instance of plugin.
And this is logical, because Ardour uses test signal to catch frequency response
in analysis, and with only one instance of the plugin this signal will go to the tails of reverberation
and other long filters. So they use separate instance with the same parameters
to do the analysis.

And this causes the problem. In my code pointers to the convolvers was global
variables. This caused interference between two plugin instances and broken pointers
as a result.

I moved pointers to the plugin class. This fixed crash problem.

But here comes the next problem. Ardour doesn't send current profile name
to this second "analysis" instance. It always created with default profile.
To avoid this I created global variable with current profile file name. I don't like this,
but I don't see any other solution by now.

As a result, plugin analysis now works, with the right profile parameters.
But! If you change profile in "Generic Controls" window, not in the own GUI
of the plugin, Ardour does not send any notification to "analysis" instance about
change. And you see not right frequency response.

If you change profile in the GUI and then open Generic Controls and plugin analysis,
all works fine. You will see frequency response of current profile.
